### ISSUE TITLE
Auto completion drawing issue

### ISSUE THREAD

**Issue by GoogleCodeExporter** [tempelmann/custom-editfield#9]
```
CEF bug: wenn man "som" tippt, erscheint "e…" an falscher Stelle - auch, wenn 
man dann Return drückt!

What steps will reproduce the problem?
1. Run the project, make sure "Enable autocomplete" as well as "Indent visually 
only" are checked.
2. Go to a line that shows an indented "end"
3.  Move the cursor past the "end", type a blank, then "som".

What is the expected output? What do you see instead?
The "e.." should appear right behind "som".

Please use labels and text to provide additional information.
The "e.." appears somewhere else, e.g. underneath "som" or futher out.

The issue does not occur if "Indent visually only" is off.

Obviously autocompletion drawing does not consider the visual indentation 
offset correctly.

```

Original issue reported on code.google.com by `tempelmann@gmail.com` on 18 Mar 2014 at 11:39

**Comment by rogermeier** [comment]
v1.7.8 still appears to have this issue. I attached a screenshot.
![screen shot 2015-03-23 at 11 14 17 pm](https://cloud.githubusercontent.com/assets/194280/6796558/60019bc0-d1b2-11e4-84ab-2846667694db.png)

**Comment by tempelmann** [comment]
@rogermeier, I don't use this feature so I am not eager to fix this. It is open source. If you find out what's going wrong exactly, let me know.
Also, the bug report suggests that this only happens with indented code, while your screenshot shows a line without indentation, so it might even have a different cause. For instance, do you build for Carbon or Cocoa? I can imagine that Cocoa has other font metrics and that the way CEF measures them is incompatible. I remember having had troubles with font width measurement in Cocoa before.

**Comment by rogermeier** [comment]
I think you're right in this may be another issue altogether. This issue is not platform dependent as I'm also seeing this on Win7:
![clipboard_2015-03-24_09-32-51](https://cloud.githubusercontent.com/assets/194280/6806917/d1439c92-d208-11e4-8940-e0873f027222.png)
What's interesting is that it only occurs when typing keys. For example, the above image shows the screen after typing "someoth". However, when using the backspace key, the autocomplete text is properly aligned. The image below shows the same screen after deleting the "h" with backspace:
![clipboard_2015-03-24_09-35-56](https://cloud.githubusercontent.com/assets/194280/6807023/38d6f494-d209-11e4-9154-a58d622da01e.png)
It behaves identically on Mac and Windows.
I'll dig through the code to see if I can find anything.

**Comment by rogermeier** [comment]
I think I'm stuck. Here is what I was able to find:
- drawContents uses the AutocompleteSuggestionInsertionX property to determine where do draw the autocomple suggestion. When a key is typed, this property is seemingly not updated to reflect the just added character.
- AutocompleteSuggestionInsertionX is updated in the AutoCompleteEOL method, which executes as a result of the keydown event, prior to drawContents.
- AutoCompleteEOL calls XYAtCharPos to update AutocompleteSuggestionInsertionX.
- XYAtCharPos calls XposForOffset for the current line to calculate the x position.
- XposForOffset the calls TextLine.TextWidth, which is where the issue appears to happen. If I inspect the text variable in TextLine.TextWidth it does not contain the typed character yet.

It appears that, at the time AutoCompleteEOL is executed after a new character is typed, the TextLine that is used to calculate where the autocomplete suggestion needs to go doesn't contain the new character yet. However, I have not been able to figure why that is exactly. 

The call stack is as follows when a new character is pressed:
CustomEditField.AutoCompleteEOL
CustomEditField.changeSelection
CustomEditField.private_replace
CustomEditField.Insert
CustomEditField.HandleKeyDown
CustomEditField.Event.KeyDown

And here is the call stack when backspace is pressed (and autocomplete works):
CustomEditField.AutoCompleteEOL
CustomEditField.changeSelection
CustomEditField.private_remove
CustomEditField.DeleteChars
CustomEditField.HandleKeyDown
CustomEditField.Event.KeyDown

If I had to guess, I would say the difference is somewhere in either .private_replace or .insert, but I can't figure out where.

**Comment by tempelmann** [comment]
I think I've found the issue. TextLine maintains a "words" array as a kind of cache, and that needs a refresh at an earlier stage. I'm trying to fix this now

**Comment by rogermeier** [comment]
Awesome! Thanks.

**Comment by tempelmann** [comment]
Dammit. Now that I've moved the comments here to a new Issue (#10), the commit comment refers to the wrong (this) issue.

**Comment by tempelmann** [comment]
Actually, this issue got closed along with issue #10 by the commit, after all!
